src/main/java/com/github/davidmoten/rtree2/geometry/internal/PolygonDouble.java

```diff
@@ -0,0 +1,142 @@
+package com.github.davidmoten.rtree2.geometry.internal;
+
+import com.github.davidmoten.rtree2.geometry.*;
+import com.github.davidmoten.rtree2.internal.util.ObjectsHelper;
+
+import java.util.ArrayList;
+
+/**
+ * A convex polygon shell.
+ */
+public final class PolygonDouble implements Polygon {
+
+    private final ArrayList<PointDouble> points;
+    private final Rectangle mbr;
+
+    private static final double PRECISION = 0.00000001;
+
+    private PolygonDouble(double[] coordinates) {
+        if (coordinates.length % 2 != 0 || coordinates.length < 6)
+            throw new IllegalArgumentException("expecting an even number of coordinate points of at least 6");
+
+        points = new ArrayList<>();
+        double minX = coordinates[0];
+        double maxX = coordinates[0];
+        double minY = coordinates[1];
+        double maxY = coordinates[1];
+        PointDouble lastPoint = PointDouble.create(coordinates[0], coordinates[1]);
+        points.add(lastPoint);
+        for (int i = 2; i < coordinates.length - 1; i += 2) {
+            // be nice to users and just ignore duplicate points
+            if (coordinates[i] == lastPoint.x() && coordinates[i + 1] == lastPoint.y()) continue;
+            minX = Math.min(minX, coordinates[i]);
+            maxX = Math.max(maxX, coordinates[i]);
+            minY = Math.min(minY, coordinates[i + 1]);
+            maxY = Math.max(maxY, coordinates[i + 1]);
+            lastPoint = PointDouble.create(coordinates[i], coordinates[i + 1]);
+            points.add(lastPoint);
+        }
+        // don't break if the polygon was closed
+        if (points.get(0).equals(lastPoint))
+            points.remove(points.size() - 1);
+        // final check to make sure the polygon remains valid after removing duplicates
+        if (points.size() < 3) throw new IllegalArgumentException("fewer than 3 unique polygon points provided");
+        mbr = RectangleDouble.create(minX, minY, maxX, maxY);
+    }
+
+    /**
+     * Constructor for a convex polygon shell.
```

Comment:
Is it a requirement that the polygon be convex? That's a significant extra requirement. We can't really call this a PolygonDouble geometry if it only supports convex polygons. ConvexPolygonDouble? I don't like that either because we  would want to deprecate it once we'd figured out how to handle concave polygons. Do you know what's involved in supporting concave polygons as well?

Reply:
To check any polygon, we would extend the point to the edge of the polygon bounding box. Whether the number of polygon edges this intersects with is odd or even determines whether it is inside. The issue is that there are a bunch of corner cases as described in the link below. If it ended up being relatively slower, it may be good to have a convex version of the class. Another option could be to check the orientation of the consecutive points similar to the point in polygon check and throw an unimplemented related exception if the direction changes (not convex).
https://www.geeksforgeeks.org/how-to-check-if-a-given-point-lies-inside-a-polygon/

Reply:
another solution strikes me that reuses your current logic is to travel along the vertices and draw straight lines from a convex vertex to the next convex vertex (skip concave vertices).  When you skip one or more concave vertices you have a another convex polygon that you will check for NON-intersection. So overall intersection is true if and only if the convex polygon with skipped vertices intersects object AND NONE of the convex polygons using the skipped vertices as described intersects the object. Might also have to be careful with intersections with the boundary of the original polygon that we want to be accepted. In the simple case where the input is a convex polygon you would make a check that all vertices are convex (we should do that regardless) and the logic is the same as you have already (there are no non-containment polygons to check).

Reply:
I don't think I follow. A diagram / drawing might help.

Reply:
Here's a diagram. B and C are polygons created by skipping concave vertices.  Object O will intersect with A iff O intersects with A + B + C (convex) AND not with B nor C. Edges might need special handling.

![image](https://user-images.githubusercontent.com/318187/227419628-38a6bb9c-3b9c-4cf8-96a3-4b99d42a6d29.png)

Reply:
Ah I realize it can get pathological. Joining two convex vertices can give you another concave vertex so you might need to back and forward track. Might need to go back to your original idea.

Reply:
In this vien (breaking a concave polygon into convex ones): https://stackoverflow.com/questions/2457840/breaking-a-concave-polygon-into-convex-ones

Reply:
>Another option could be to check the orientation of the consecutive points similar to the point in polygon check and throw an unimplemented related exception if the direction changes (not convex).
https://www.geeksforgeeks.org/how-to-check-if-a-given-point-lies-inside-a-polygon/

I'm open to this too. We can state in the javadoc for PolygonDouble that it only supports convex polygons for the moment (and suggest that if the user has a non-convex polygon that they split it into convex polygons themselves and make multiple calls).

Reply:
Throws UnsupportedOperationException if it is not a convex polygon, this is also documented in the javadoc.